app/Jobs/EncodeTrackFile.php

```diff
@@ -0,0 +1,131 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Poniverse\Ponyfm\Jobs;
+
+use Carbon\Carbon;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\Log;
+use OAuth2\Exception;
+use Poniverse\Ponyfm\Jobs\Job;
+use Illuminate\Queue\SerializesModels;
+use Illuminate\Queue\InteractsWithQueue;
+use Illuminate\Contracts\Bus\SelfHandling;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Poniverse\Ponyfm\Track;
+use Poniverse\Ponyfm\TrackFile;
+use Symfony\Component\Process\Exception\ProcessFailedException;
+use Symfony\Component\Process\Process;
+
+class EncodeTrackFile extends Job implements SelfHandling, ShouldQueue
+{
+    use InteractsWithQueue, SerializesModels;
+    /**
+     * @var TrackFile
+     */
+    private $trackFile;
+    /**
+     * @var
+     */
+    private $isExpirable;
+
+    /**
+     * Create a new job instance.
+     * @param TrackFile $trackFile
+     * @param $isExpirable
+     */
+    public function __construct(TrackFile $trackFile, $isExpirable)
+    {
+        $this->trackFile = $trackFile;
+        $this->isExpirable = $isExpirable;
+    }
+
+    /**
+     * Execute the job.
+     *
+     * @return void
+     */
+    public function handle()
+    {
+        // Start the job
+        $this->trackFile->in_progress = true;
+        $this->trackFile->update();
+
+        // Assign the source to the track's master file
+        $source = TrackFile::where('track_id', $this->trackFile->track_id)
+            ->where('is_master', true)
+            ->first()
+            ->getFile();
+
+        // Assign the target
+        $destination = $this->trackFile->track->getDirectory();
+        $this->trackFile->track->ensureDirectoryExists();
+        $target = $destination . '/' . $this->trackFile->getFilename();
+
+        // Prepare the command
+        $format = Track::$Formats[$this->trackFile->format];
+        $command = $format['command'];
+        $command = str_replace('{$source}', '"' . $source . '"', $command);
+        $command = str_replace('{$target}', '"' . $target . '"', $command);
+
+        Log::info('Encoding track file ' . $this->trackFile->id . ' into ' . $target);
+
+        // Start a synchronous process to encode the file
+        $process = new Process($command);
+        try {
+            $process->mustRun();
+        } catch (ProcessFailedException $e) {
+            Log::error('An exception occured in the encoding process for track file ' . $this->trackFile->id . ' - ' . $e->getMessage());
+            // Ensure queue fails
+            throw $e;
+        } finally {
+            Log::info($process->getOutput());
+        }
+
+        // Update the tags of the track
+        $this->trackFile->track->updateTags();
```

Comment:
This will attempt to update the tags for all of the track's files, even ones that don't exist at the moment. Look into refactoring the `updateTags()` method to be in the `TrackFile` class, and to gracefully handle being called on a cacheable file that doesn't exist at the moment.